File: app/src/main/java/com/concrete/android/challenge/ui/pull_request/PullRequestAdapter.java
package com.concrete.android.challenge.ui.pull_request;

import android.view.LayoutInflater;
import android.view.ViewGroup;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.concrete.android.challenge.data.model.PullRequest;
import com.concrete.android.challenge.databinding.ItemPullRequestBinding;
import com.concrete.android.challenge.ui.base.BaseViewHolder;
import com.concrete.android.challenge.ui.base.BindableAdapter;
import java.util.List;

/**
 * @author Thiago Corredo
 * @since 2019-07-12
 */
public class PullRequestAdapter extends RecyclerView.Adapter<BaseViewHolder> implements
    BindableAdapter<PullRequest> {

  private List<PullRequest> pullRequests;

  private PullRequestItemViewModel.PullRequestItemViewModelListener listener;

  public PullRequestAdapter() {
  }

  public PullRequestAdapter(List<PullRequest> pullRequests) {
    this.pullRequests = pullRequests;
  }

  @NonNull @Override
  public BaseViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    ItemPullRequestBinding itemPullRequestBinding =
        ItemPullRequestBinding.inflate(LayoutInflater.from(parent.getContext()),
            parent, false);
    return new PullRequestViewHolder(itemPullRequestBinding);
  }

  @Override public void onBindViewHolder(@NonNull BaseViewHolder holder, int position) {
    holder.onBind(position);
  }

  @Override public int getItemCount() {
    return pullRequests.size();
  }

  public void add(List<PullRequest> pullRequests) {
    this.pullRequests.addAll(pullRequests);
    notifyDataSetChanged();
  }

  public void clearItems() {
    pullRequests.clear();
    notifyDataSetChanged();
  }

  @Override public void setItems(List<PullRequest> pullRequests) {
    this.pullRequests.addAll(pullRequests);
    notifyDataSetChanged();
  }

  public void setListener(
      PullRequestItemViewModel.PullRequestItemViewModelListener listener) {
    this.listener = listener;
  }

  public class PullRequestViewHolder extends BaseViewHolder {

    private ItemPullRequestBinding itemPullRequestBinding;

    private PullRequestItemViewModel itemViewModel;

    public PullRequestViewHolder(ItemPullRequestBinding itemPullRequestBinding) {
      super(itemPullRequestBinding.getRoot());
      this.itemPullRequestBinding = itemPullRequestBinding;
    }

    @Override public void onBind(int position) {
      PullRequest pullRequest = pullRequests.get(position);
      itemViewModel = new PullRequestItemViewModel(pullRequest, listener);
      itemPullRequestBinding.setViewModel(itemViewModel);
      itemPullRequestBinding.executePendingBindings();
    }
  }
}

File: app/src/main/java/com/concrete/challenge/ui/adapters/PullRequestAdapter.kt
package com.concrete.challenge.ui.adapters

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.concrete.challenge.R
import com.concrete.challenge.data.PullRequestEntity
import com.concrete.challenge.ui.viewholders.PullRequestViewHolder

class PullRequestAdapter(
    private val manager: AdapterManager
) : RecyclerView.Adapter<PullRequestViewHolder>() {

    interface AdapterManager {
        fun onPullRequestClicked(pullRequestClicked: PullRequestEntity)
    }

    private val pullRequestList = mutableListOf<PullRequestEntity>()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PullRequestViewHolder {
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.item_pull_request, parent, false)

        return PullRequestViewHolder(view, manager)
    }

    override fun getItemCount(): Int = pullRequestList.size

    override fun onBindViewHolder(holder: PullRequestViewHolder, position: Int) {
        val item = pullRequestList[position]

        holder.bind(item)
    }

    fun addItems(list: List<PullRequestEntity>) {
        pullRequestList.addAll(list)
        notifyDataSetChanged()
    }

}